modules/caddytls/capools.go

```diff
@@ -0,0 +1,803 @@
+package caddytls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"reflect"
+
+	"github.com/caddyserver/certmagic"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/modules/caddypki"
+)
+
+func init() {
+	caddy.RegisterModule(InlineCAPool{})
+	caddy.RegisterModule(FileCAPool{})
+	caddy.RegisterModule(PKIRootCAPool{})
+	caddy.RegisterModule(PKIIntermediateCAPool{})
+	caddy.RegisterModule(StoragePool{})
+	caddy.RegisterModule(LazyCertPool{})
+}
+
+// The interface to be implemented by all guest modules part of
+// the namespace 'tls.ca_pool.source.'
+type CA interface {
+	CertPool() *x509.CertPool
+}
+
+// InlineCAPool is a certificate authority pool provider coming from
+// a DER-encoded certificates in the config
+type InlineCAPool struct {
+	// A list of base64 DER-encoded CA certificates
+	// against which to validate client certificates.
+	// Client certs which are not signed by any of
+	// these CAs will be rejected.
+	TrustedCACerts []string `json:"trusted_ca_certs,omitempty"`
+
+	pool *x509.CertPool
+}
+
+// CaddyModule implements caddy.Module.
+func (icp InlineCAPool) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID: "tls.ca_pool.source.inline",
+		New: func() caddy.Module {
+			return new(InlineCAPool)
+		},
+	}
+}
+
+// Provision implements caddy.Provisioner.
+func (icp *InlineCAPool) Provision(ctx caddy.Context) error {
+	caPool := x509.NewCertPool()
+	for i, clientCAString := range icp.TrustedCACerts {
+		clientCA, err := decodeBase64DERCert(clientCAString)
+		if err != nil {
+			return fmt.Errorf("parsing certificate at index %d: %v", i, err)
+		}
+		caPool.AddCert(clientCA)
+	}
+	icp.pool = caPool
+
+	return nil
+}
+
+// Syntax:
+//
+//	trust_pool	inline {
+//		trust_der <base64_der_cert>...
+//	}
+//
+// The 'trust_der' directive can be specified multiple times.
+func (icp *InlineCAPool) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	for d.Next() {
+		if d.CountRemainingArgs() > 0 {
+			return d.ArgErr()
+		}
+		for nesting := d.Nesting(); d.NextBlock(nesting); {
+			switch d.Val() {
+			case "trust_der":
+				icp.TrustedCACerts = append(icp.TrustedCACerts, d.RemainingArgs()...)
+			default:
+				return fmt.Errorf("unrecognized directive: %s", d.Val())
+			}
+		}
+	}
```

Comment:
See https://github.com/caddyserver/caddy/pull/6037 :sweat_smile: 

```suggestion
	d.Next() // consume module name
	if d.CountRemainingArgs() > 0 {
		return d.ArgErr()
	}
	for nesting := d.Nesting(); d.NextBlock(nesting); {
		switch d.Val() {
		case "trust_der":
			icp.TrustedCACerts = append(icp.TrustedCACerts, d.RemainingArgs()...)
		default:
			return fmt.Errorf("unrecognized directive: %s", d.Val())
		}
	}
```